### PR TITLE
JAPI-488 Illegal capacity error reading child dataset field in index

### PR DESCRIPTION
**Pull request by jpmcmu**
- Corrected issue with how biased integer correction for indexes was implemented.

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [X] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana @drealeed please review

File: commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java [hunk 1]
```diff
@@ -933,8 +933,12 @@ private static FieldDef parseJsonTypeDefinition(JSONObject jsonTypeDefinitions,
             }
             default:
             {
-                return new FieldDef("", fieldType, fieldType.description(), length, isFixedLength(typeID), isUnsigned(typeID), getSourceType(typeID),
-                        new FieldDef[0]);
+                FieldDef fd = new FieldDef("", fieldType, fieldType.description(), length,
+                        isFixedLength(typeID), isUnsigned(typeID), getSourceType(typeID), new FieldDef[0]);
+                if ((typeID & TYPE_ID_MASK) == type_keyedint) {
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java:938]: formatting

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 2]
```diff
@@ -669,11 +669,13 @@ private void readIntoScratchBuffer(int offset, int dataLen) throws IOException
...
      * @param little_endian
      *            true if the value is little endian
+     * @param shouldCorrectBias 
+     *            true if the value should be corrected for index bias
      * @return the integer extracted as a long
      * @throws IOException
      *             Signals that an I/O exception has occurred.
-     */
-    private long getInt(int len, boolean little_endian) throws IOException
+     */
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:677]: minor, superfluous white-space introduced

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu seems fine, only a couple of minor issues

**Comment by jpmcmu** [comment]
@rpastrana please review

File: dfsclient/src/test/resources/child-index.ecl [hunk 3]
```diff
@@ -0,0 +1,25 @@
+childRec := RECORD
```
Comment (rpastrana) [inline dfsclient/src/test/resources/child-index.ecl:1]: at the risk of being unnecessarily verbose, let's ensure the file name is self documenting...
generate-nested-dataset-and-index.ecl?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 4]
```diff
@@ -854,6 +862,33 @@ private void writeFile(List<HPCCRecord> records, String fileName, FieldDef recor
...
+        wu.setJobname("IndexCreation " + indexName);
+        wu.setCluster(thorclustername);
+
+        HPCCWsWorkUnitsClient client = wsclient.getWsWorkunitsClient();
+        String result = client.createAndRunWUFromECLAndGetResults(wu);
+
+        HPCCFile file = new HPCCFile(indexName, connString, hpccUser, hpccPass);
+        try
+        {
+            List<HPCCRecord> records = readFile(file, connTO, false);
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:888]: does it make sense to assert an expected number of records?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 5]
```diff
@@ -854,6 +862,33 @@ private void writeFile(List<HPCCRecord> records, String fileName, FieldDef recor
         }
     }
 
+    @Test
+    public void readIndexTest() throws Exception
+    {
+        String indexName = "test::index::child_dataset::key";
+
+        String eclFile = "child-index.ecl";
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:870]: let's externalize this functionality, maybe a function in BaseRemoteTest which loads and execute the ecl based on file name. we could execute the appropriate ecl once before the test run, and the tests should be skipped if the required datasets are not available.

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu a few comments on the test functionality

**Review by rpastrana — APPROVED** [review]

**Pull request merged**